### DISCUSSION TITLE
How to provoke an error and success response on the same GET endpoint

### DISCUSSION THREAD

**Discussion by kevinmamaqi** [mswjs/msw#783] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I use the library for testing purposes with jest and @testing-library/react. When I do a fetch request to a GET endpoint, how can I return once an error and once a success to test both states of my app?

**Describe the solution you'd like**
A flag.

**Describe alternatives you've considered**
Sending an extra parameter in the request to filter it in the endpoint.

**Comment by kettanaito** [comment, marked as answer]
Hey, @kevinmamaqi. You can use a [runtime request handler](https://mswjs.io/docs/api/setup-server/use) to override which response is being returned for the same endpoint to test multiple states of your code.

```js
import { rest } from 'msw'
import { setupServer } from 'msw/node'

const server = setupServer(
  // I recommend listing the success scenarios when declaring the server.
  rest.get('/resource', (req, res, ctx) => {
    return res(ctx.json({ id: 1 }))
  })
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

it('handles a successful response', () => {
  // run your code (i.e. render components, execute functions)
  // without any additions, this test case will receive a successful response
  // defined above.
})

it('handles an error response', () => {
  server.use(
    // In this paticular test respond to the same request with a 404 response.
    rest.get('/resource', (req, res, ctx) => {
      return res(ctx.status(404))
    })
  )

  // run your code here.
  // within this test a "GET /resource" will receive a 404 error response
  // specified above.
})
```

Don't forget to call [.resetHandlers()](https://mswjs.io/docs/api/setup-server/reset-handlers) after each test so that the `.use()` overrides you add don't affect irrelevant tests.
